README.md
# SASxport - Read and Write SAS XPORT Files

[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Linux Build Status](https://travis-ci.org/warnes/SASxport.svg?branch=master)](https://travis-ci.org/warnes/SASxport)
[![Windows Build status](https://ci.appveyor.com/api/projects/status/github/warnes/SASxport?svg=true&passingText=Windows%20build%20-%20passing&pendingText=Windows%20build%20-%20pending&failingText=Windows%20build%20-%20failing)](https://ci.appveyor.com/project/warnes/SASxport)
[![](http://www.r-pkg.org/badges/version/SASxport)](http://www.r-pkg.org/pkg/SASxport)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/SASxport)](http://www.r-pkg.org/pkg/SASxport)
[![Coverage Status](https://codecov.io/gh/warnes/SASxport/branch/master/graph/badge.svg)](https://codecov.io/gh/warnes/SASxport)

<!--- The current version of Pandoc used with RStudio doesn't handle this properly.  Try again in the next version. ---!>
<!--- [![Linux Build Status](https://img.shields.io/travis/warnes/SASxport.svg?label=Linux)](https://travis-ci.org/warnes/SASxport)
--->

This package provides functions for reading, listing
the contents of, and writing SAS xport format files.
The functions support reading and writing of either
individual data frames or sets of data frames.  Further,
a mechanism has been provided for customizing how
variables of different data types are stored.

## Installation

### From CRAN

```r
install.packages("SASxport")
```

### From GitHub 

```r
devtools::install_github("warnes/SASxport")
```

## Usage

```r
library(SASxport)
```

### Writing SAS XPORT files

`write.xport` writes one or more data frames into a SAS XPORT format
library file.

The function creates a SAS XPORT data file (see reference) from
one or more data frames.  This file format imposes a number of
constraints:
* Data set and variable names are truncated to 8 characters and
  converted to upper case.  All characters outside of the set
  A-Z, 0-9, and '\_' are converted to '\_'.
* Character variables are stored as characters.
* If `autogen.formats = TRUE` (the default), factor variables are
  stored as numeric with an appropriate SAS format
  specification. If `autogen.formats = FALSE`, factor variables
  are stored as characters.
* All numeric variables are stored as double-precision floating
  point values utilizing the IBM mainframe double precision
  floating point format (see the reference).
* Date and time variables are either converted to number of
  days since 1960-01-01 (date only), or number of seconds since
  1960-01-01:00:00:00 GMT (date-time variables).
* Missing values are converted to the standard SAS missing
  value `.`

The SAS XPORT format allows each dataset to have a label and a
type (set via the `label` and `SAStype` functions).  In addition,
each variable may have a corresponding label, display format, and
input format.  To set these values, add the attribute `label`,
`SASformat`, or `SASiformat` to individual data frame.  These
attributes may be set using the `label`, `SASformat`, and
`SASiformat` functions. (See examples in the package.)

The actual translation of R objects to objects appropriate for SAS
is handled by the `toSAS` generic and associated methods, which
can be (re)defined by the user to provide fine-grained control.

### Reading SAS XPORT files

`read.xport` reads a file as a SAS XPORT format library and returns a list
of data frames:
* SAS date, time, and date/time variables are converted
  respectively to `Date`, POSIX, or `chron` objects.
* SAS labels are stored in `label` attributes on each variable,
  and are accessible using the `label` function.
* SAS formats are stored in `SASformat` attributes on each
  variable, and are accessable using `SASformat`.
* SAS iformats are stored in `SASiformat` attributes on each
  variable, and are accessable using `SASiformat`.
* SAS integer variables are stored as integers unless
  `force.integer` is `FALSE`.

If the file includes the output of `PROC FORMAT CNTLOUT=`,
variables having customized label formats will be converted to
`factor` objects with appropriate labels.

If a datasets in the original file has a label or type, these will
be stored in the corresponding `label` and `SAStype` attributes,
which can be accessed by the `label` and `SAStype` functions.

## License

GPL 2.1 © Gregory R. Warnes, United States Government, Frank E. Harrell,
          Jr., Douglas M. Bates, Mango Solutions

inst/NEWS.md
Version 1.5.4 - 2018-01-29
--------------------------

Bug fixes:

- Methods associated with "Hmisc::label" are no longer imported explicitly
- Code flow improved in places
- Maintainer details updated: Owen Jones <ojones@mango-solutions.com>


Version 1.5.0 - 2014-07-21
--------------------------

Bug fixes:

- Now works properly on big-endian systems such as the PowerPC,
  Spark. (Reported by Brian Ripley )x

- Explicitly cast left bit shifts to avoid undefined C language
  behavior. (Reported by Brian Ripley)

- Resolve problem in accessing "Hmisc::label.default<-" if
  SASxport::read.xport is called without loading SASxport.  (Reported
  by Dominic Comtois)

- 'read.xport' now preserves '$' at the beginning of SAS character
  format and iformat strings.  (Reported by Dominic Comtois)

- 'read.xport' argument names.tolower was not being honored for
   dataset names.  (Reported by Dominic Comtois)

Other changes:

- Modified several test files to display generated .xpt data so that
  issues can be more easily detected and diagnosed.

- C code cleanup and reorgaization to improve clarity.


Version 1.4.0 - 2014-04-09
--------------------------

API Change:

- SASxport now relies on the 'label' methods defined by the Hmisc
  package instead of defining its own.

Bug fixes:

- The 'read.xport' and 'write.xport' functions were failing when both
  the SASxport and the Hmisc packages were loaded due to conflicts
  between the label methods defined by each package.  This has been
  resolved by removing the label methods from SASxport and using those
  from Hmisc instead.


Version 1.3.6 2013-10-09
------------------------

Bug fixes:

- In manual pages for read.xport() and lookup.xporT(): Update URL for
  'test2.xpt', and use a local copy for executed example code.

Version 1.3.5 2013-06-14
------------------------

Bug fixes:

- read.xport() and write.xport() now properly handle empty
  dataset/dataframe objects.

Version 1.3.4 2013-05-31
------------------------

Bug fixes:

- Correct error in write.xport when a factor contains only NA entries.

Other Changes:

- Package test scripts now use a fixed timezone to prevent unhelpful warnings.

Version 1.3.2 2013-05-11
------------------------

New features:

- dataset label and type are now supported.  See write.xport() and
  read.xport() for examples.

Bug fixes:

- Integrate patch from foreign package to properly handle xport files
  with datasets that end exactly on an 80-byte record boundry & add
  corresponding test file.

- Replace file.path(path.package(...)) with system.file(...)

Changes:

- Remove oboslete .First.lib() function

- Replace file.path(path.package(...)) with system.file(...)

Version 1.3.1 2013-03-24
------------------------

Changes:

- Replace use of depreciated .path.package() with path.package() for R 3.0.0.

Version 1.3.0 2012-06-27
------------------------

New features:

- New function makeSASNames() to create valid and unique SAS names
  from character vectors.

Bug fixes:

- Improper handling of duplicates names in write.xport() was
  generating names longer than 8 characters, resulting in invalid
  files.  Corrected by using the new makeSASNames() function instead
  of the R make.names() function.


Version 1.2.4 2010-11-11
------------------------

Bug fixes:

- Fix bug in handling of 'as.is' argument to read.xport ('as.is=TRUE'
  was not operating as documented).


Version 1.2.3 2008-02-29
------------------------

Bug fixes:

- Fix typo in manual page for write.xport() reported by Yinlai Meng.


Version 1.2.2 2007-11-09
------------------------

Bug fixes:

-  Apply patches to fix problems on 64 bit platforms, as sumitted by
   Brian Ripley.


Version 1.2.1 2007-11-05
-------------------------

Other:

- Correct warning message due to extraneous ';' charcters after
  function closing braces.


Version 1.2.0 2007-11-01
-------------------------

New Features:

- SAS format and iformat information is now accessed via 'SASformat()'
  and 'SASiformat()' functions instead of 'formats' and 'iformat'.  The
  information accessed by these functions is now stored in attributes
  with the same name.  This avoids conflicts with the use of 'format'
  by chron objects.

- Copies of the code for foreign::read.xport and foreign::lookup.xport
  is now part of the SASxport package, permitting extension to these
  functions as needed, and removing the dependency on the foreign
  package.

- Overflow of SASxport numeric format values, which have a smaller
  range than IEEE 754 numeric values now standard,  now generates NAN
  instead of 0.0.


Bug Fixes:

- Fix for problem storing negative numbers.

- SAS format length and digit information is now properly captured
  by read.xport().  This is supported by an improved version of
  lookup.xport().

- SAS format information was not being properly utilized when more
  than one format was present.

- Improved handling of SAS date formats


Other:

- Test routines added to test handling of numeric values.


Version 1.1.1
-------------------------

-  Display support information at package startup


Version 1.1.0 -
-------------------------

New Features:

- Add support for autogeneration of SAS FORMAT information as a PROC
  CONTENTS fmtin= dataset.   This enables R factors to be handled
  properly on the receiving system.

Bug Fixes:


Other:



Version 1.0.0
-------------

New Fetures:



Bug Fixes:



Other:




Beta 3 - 2007-08-10
-------------------

New features:

- read.xport's names.tolower argument now defaults to FALSE so that
  variable (and data set) names are now left as uppercase.

- Improved crediting of BRL-CAD source code

Bug fixes:

- Correct call to sprintf where printf was intended in src/ieee2ibm.c

Other:

- Augmented ieee2ibm code with corresponding ibm2ieee code for
  completeness.


Beta 2 - 2007-08-10
-------------------

New Features:

- Replaced IEEE to IBM translation code with GPL'ed version from BPL-CAD.

Bug Fixes:


R/write.xport.R


#' @importFrom Hmisc label
#' @export

write.xport <- function(...,
                        list = base::list(),
                        file = stop("'file' must be specified"),
                        verbose = FALSE,
                        sasVer = "7.00",
                        osType,
                        cDate = Sys.time(),
                        formats = NULL,
                        autogen.formats = TRUE) 
    {
    ## Handle verbose option ##
    oldDebug <- getOption("DEBUG")

    options(DEBUG = verbose)
    
    on.exit(options(DEBUG = oldDebug))
    
    ## Handle osType default value ##
    if (missing(osType)) {
        osType <- paste("R ", R.version$major, ".", R.version$minor, sep = "")
    }
    
    ## Handle '...' ##
    dotList <- base::list(...)
    dotNames <- names(dotList)
    
    if (is.null(dotNames)) {
        dotNames <- rep("", length(dotList))
    }
    
    if (length(dotList) > 0) {
        ## Get data frame names from ... in function call, but don't
        ## clobber any explicitly provided names
        mc <- match.call()
        mc$file <- NULL
        mc$verbose <- NULL
        mc$sasVer <- NULL
        mc$osType <- NULL
        mc$cDate <- NULL
        mc$list <- NULL
        mc$autogen.formats <- NULL
        mc[[1]] <- NULL
        # note we *do not* mask off format argument so it will get
        # magically included if present.
        
        mc <- as.character(mc)
        
        badNames <- which(is.na(dotNames) | dotNames <= "")
        dotNames[badNames] <- mc[badNames]
    }
    
    ## Join explicit 'list' argument to '...' arguments ##
    listNames <- names(list)
    if (is.null(listNames)) {
        listNames <- rep("", length(list))
    }
    dfList  <- c(dotList, list)
    dfNames <- c(dotNames, listNames)
    
    ## check for and handle <NA> or empty names ##
    badNames <- which(is.na(dfNames) | dfNames <= "")
    if (length(badNames) > 0) {
        warning("Replacing missing or invalid dataset names")
        dfNames[badNames] = paste("DATA", badNames, sep = "")
    }
    
    ## put revised names back ##
    names(dfList) <- dfNames
    
    #######
    ##
    scat("Ensure all objects to be stored are data.frames...\n")
    not.df <- which(!sapply(dfList, is.data.frame))
    if (any(not.df)) {
        if (length(not.df) == 1) {
            stop(paste("'", dfNames[not.df], "'"),
                 " is not a data.frame object.")
        } else {
            stop(paste("'", dfNames[not.df], "'", sep = "", collapse = ", "),
                 " are not data.frame objects.")
        }
    }
    ##
    #######
    
    
    #######
    ##
    scat("Ensure object names are valid and unique...\n")
    names(dfList) <- dfNames <- makeSASNames(dfNames)
    ##
    #######
    
    #######
    ## Generate formats for factor variables
    if (autogen.formats) {
        dfList <- make.formats(dfList, formats = formats)
        dfNames <- names(dfList)
        formats <- dfList$FORMATS
    }
    
    if (is.null(formats) || length(formats) < 1 || nrow(formats) < 1) {
        formats <- NULL
        dfList$FORMATS <- NULL
        dfNames <- names(dfList)
    }
    
    #######
    scat("opening file ...")
    if (is.character(file)) {
        if (file == "") {
            file <- stdout()
            
        } else {
            file <- file(description = file, open = "wb")
            on.exit(close(file))
        }
    }
    scat("Done")
    
    if (file == stdout()) {
        out <- function(...) {
            cat("ASCII: ", rawToDisplay(...), "\n")
            cat("HEX:   ", ..., "\n")
        }
    } else {
        out <- function(...) {
            writeBin(..., raw(), con = file)
        }
    }
    
    scat("Write file header ...")
    out(xport.file.header(
        cDate = cDate,
        sasVer = sasVer,
        osType = osType
    ))
    scat("Done.")
    
    for (i in dfNames) {
        df <- dfList[[i]]
        
        if (is.null(colnames(df))) {
            colnames(df) <- rep("", length = ncol(df))
        }
        
        emptyFlag <- (colnames(df) == "" | is.na(colnames(df)))
        
        if (any(emptyFlag)) {
            warning("Unnamed variables detected, using default names")
            colnames(df)[emptyFlag] = paste("VAR", 1:length(emptyFlag), sep = "")
            dfList[[i]] <- df
        }
        
        
        colnames(dfList[[i]]) <-
            colnames(df) <- varNames <- makeSASNames(colnames(df))
        
        offsetTable <- data.frame(
            "name" = varNames,
            "len" = rep(NA, length(varNames)),
            "offset" = rep(NA, length(varNames))
        )
        rownames(offsetTable) <- offsetTable[, "name"]
        
        dfLabel <- label(df, default = "", self = TRUE)
        dfType <- SAStype(df, default = "")
        
        scat("Write data frame header ...")
        out(
            xport.member.header(
                dfName = i,
                cDate = cDate,
                sasVer = sasVer,
                osType = osType,
                dfLabel = dfLabel,
                dfType = dfType
            )
        )
        scat("Done.")
        
        scat("Write variable information block header ...")
        out(xport.namestr.header(nvar = ncol(df)))
        scat("Done.")
        
        scat("Write entries for variable information block ...")
        lenIndex <- 0
        varIndex <- 1
        spaceUsed <- 0
        for (i in colnames(dfList[[i]])) {
            scat("", i , "...")
            var <- df[[i]]
            
            # get attribute information before any transformations!"
            varLabel <- attr(var, "label")
            varFormat <- attr(var, "SASformat")
            varIFormat <- attr(var, "SASiformat")
            
            # Convert R object to SAS object
            df[[i]] <- var <- toSAS(var, format.info = formats)
            
            # compute variable length
            # From R 3.3.0 NA is returned by nchar if the
            # the argument is NA, unless keepNA = FALSE is supplied
            # In older versions nchar(NA) is 2, and this is the
            # behavior we need now
            if (is.character(var)) {
                if ("keepNA" %in% names(as.list(args(nchar)))) {
                    varLen <- max(c(8, nchar(var, keepNA = FALSE)))
                } else {
                    varLen <- max(c(8, nchar(var)))
                }
            } else {
                varLen <- 8
            }
            
            # fill in variable offset and length information
            offsetTable[i, "len"]    <- varLen
            offsetTable[i, "offset"] <- lenIndex
            
            # parse format and iformat
            formatInfo  <- parseFormat(varFormat)
            iFormatInfo <- parseFormat(varIFormat)
            
            # write the entry
            out(
                xport.namestr(
                    var = var,
                    varName   = i,
                    varNum    = varIndex,
                    varPos    = lenIndex,
                    varLength = varLen,
                    varLabel  = varLabel,
                    fName   = formatInfo$name,
                    fLength = formatInfo$len,
                    fDigits = formatInfo$digits,
                    iName   = iFormatInfo$name,
                    iLength = iFormatInfo$len,
                    iDigits = iFormatInfo$digits
                )
            )
            
            # increment our counters
            lenIndex <- lenIndex + varLen
            varIndex <- varIndex + 1
            spaceUsed <- spaceUsed + 140
        }
        scat("Done.")
        
        # Space-fill to 80 character record end
        fillSize <- 80 - (spaceUsed %% 80)
        if (fillSize == 80) {
            fillSize <- 0
        }
        out(xport.fill(TRUE, fillSize))
        
        scat("Write header for data block ...")
        out(xport.obs.header())
        scat("Done")
        
        scat("Write data ... ")
        
        spaceUsed <- 0
        if (nrow(df) > 0) {
            for (i in 1:nrow(df)) {
                for (j in 1:ncol(df)) {
                    val <- df[i, j]
                    valLen <- offsetTable[j, "len"]
                    
                    scat("i=",
                         i,
                         " j=",
                         j,
                         " value=",
                         val,
                         " len=",
                         valLen,
                         "")
                    
                    if (is.character(val)) {
                        out(xport.character(val, width = valLen))
                    } else {
                        out(xport.numeric(val))
                    }
                    
                    spaceUsed <- spaceUsed + valLen
                }
            }
        }
        
        fillSize <- 80 - (spaceUsed %% 80)
        if (fillSize == 80) {
            fillSize <- 0
        }
        out(xport.fill(TRUE, fillSize))
        
        scat("Done.")
    }
    
    scat("Closing file ...")
    if (is.character(file)) {
        if (file != "") {
            close(file)
            on.exit()
        }
    }
    scat("Done")
    
}


R/read.xport.R
##
## Code originally from Frank Harrell's 'Hmisc' library:
##   http://biostat.mc.vanderbilt.edu/twiki/bin/view/Main/Hmisc
## Copied with permission on 2007-08-04
##

#' @importFrom Hmisc label
#' @importFrom utils download.file
#' @export

read.xport <- function(file,
                       force.integer = TRUE,
                       formats = NULL,
                       name.chars = NULL,
                       names.tolower = FALSE,
                       keep = NULL,
                       drop = NULL,
                       as.is = 0.95, # Prevent factor conversion if 95% or more unique
                       verbose = FALSE,
                       as.list = FALSE,
                       include.formats = FALSE) {
    
    sasdateform <- toupper(
        c(
            "date",
            "mmddyy",
            "yymmdd",
            "ddmmyy",
            "yyq",
            "monyy",
            "julian",
            "qtr",
            "weekdate",
            "weekdatx",
            "weekday",
            "month"
        )
    )
    sastimeform <- toupper(c("hhmm", "hour", "mmss", "time"))
    sasdatetimeform <- toupper(c("datetime", "tod"))
    
    if (verbose) {
        oldOptions <- options("DEBUG")
        options(DEBUG = TRUE)
        on.exit(options(oldOptions))
    }
    
    if (length(grep('http://', file)) > 0 ||
        length(grep('ftp://', file)) > 0) {
        scat("Downloading file...")
        tf <- tempfile()
        download.file(file, tf, mode = 'wb', quiet = TRUE)
        file <- tf
    }
    
    scat("Checking if the specified file has the appropriate header")
    xport.file.header <-
        "HEADER RECORD*******LIBRARY HEADER RECORD!!!!!!!000000000000000000000000000000  "
    file.header <-
        readBin(
            file,
            what = character(0),
            n = 1,
            size = nchar(xport.file.header)
        )
    file.header <-
        substr(file.header,
               start = 1,
               stop = nchar(xport.file.header))
    if (!identical(xport.file.header, file.header)) {
        stop("The specified file does not start with a SAS xport file header!")
    }
    
    scat("Extracting data file information...")
    dsinfo <- lookup.xport.inner(file)
    
    dsLabels <- sapply(dsinfo, Hmisc::label)
    dsTypes  <- sapply(dsinfo, SAStype)
    
    if (length(keep)) {
        whichds <- toupper(keep)
    } else {
        whichds <-
            setdiff(names(dsinfo),
                    c(toupper(drop), '_CONTENTS_', '_contents_'))
    }
    
    scat("Reading the data file...")
    ds <- read.xport.inner(file, stringsAsFactors = FALSE)
    if (any(duplicated(names(dsinfo)))) {
        # only true if file contains has more than one data set
        warning("Duplicate data set names in file.  Data set names have been made unique.")
        names(dsinfo) <- make.unique(names(dsinfo))
        names(ds) <- make.unique(names(ds))
    }
    
    
    if ((length(keep) > 0 || length(drop) > 0)) {
        ds <- ds[whichds]
    }
    
    scat("Processing contents...")
    ## PROC FORMAT CNTLOUT= dataset present?
    fds <- NULL
    if (!length(formats)) {
        fds <- sapply(dsinfo, function(x)
            all(
                c('FMTNAME', 'START', 'END', 'MIN', 'MAX', 'FUZZ')
                %in% x$name
            ))
        fds <- names(fds)[fds]
        if (length(fds) > 1) {
            warning(
                'transport file contains more than one PROC FORMAT CNTLOUT= dataset; using only the first'
            )
            fds <- fds[1]
        }
    }
    
    finfo <- NULL
    if (length(formats) || length(fds)) {
        if (length(formats)) {
            finfo <- process.formats(formats)
        } else {
            finfo <- process.formats(ds[[fds]])
        }
    }
    
    ## Number of non-format datasets
    nods <- length(whichds)
    nds <- nods - (length(formats) == 0 && length(finfo) > 0)
    which.regular <- setdiff(whichds, fds)
    
    ## Handle lowercase name conversions
    if (names.tolower) {
        names.tolower <- tolower
    } else {
        names.tolower <- function(x)
            x
    }
    
    dsn <- names.tolower(which.regular)
    
    res <- vector('list', nds)
    names(res) <- gsub('_', '.', dsn)
    
    
    possiblyConvertChar <- (is.logical(as.is) && !as.is) ||
        (is.numeric(as.is) && as.is > 0)
    j <- 0
    for (k in which.regular) {
        j <- j + 1
        scat('Processing SAS dataset', k)
        w <- if (nods == 1) {
            ds
        } else {
            ds[[k]]
        }
        
        scat('.')
        
        Hmisc::label(w, self = TRUE)   <- dsLabels[k]
        names(Hmisc::label(w, self = TRUE)) <- NULL
        SAStype(w) <- dsTypes[k]
        names(SAStype(w)) <- NULL
        
        nam <-
            names.tolower(makeNames(names(w), allow = name.chars))
        names(w) <- nam
        dinfo <- dsinfo[[k]]
        
        fmt <- dinfo$format
        formats <- fstr(fmt, dinfo$flength, dinfo$fdigits)
        
        ifmt <- dinfo$iformat
        iformats <- fstr(ifmt, dinfo$iflength, dinfo$ifdigits)
        
        lab <- dinfo$label
        
        ndinfo <-
            names.tolower(makeNames(dinfo$name, allow = name.chars))
        names(lab) <-
            names(fmt) <-
            names(formats) <- names(iformats) <- ndinfo
        if (length(w) > 0) {
            for (i in 1:length(w)) {
                changed <- FALSE
                x  <- w[[i]]
                fi <- fmt[nam[i]]
                
                names(fi) <- NULL
                if (fi != '' &&
                    length(finfo) && (fi %in% names(finfo))) {
                    f <- finfo[[fi]]
                    if (length(f)) {
                        ## may be NULL because had a range in format
                        x <- factor(x, f$value, f$label)
                        attr(x, 'SASformat') <- fi
                        changed <- TRUE
                    }
                }
                
                if (is.numeric(x)) {
                    if (fi %in% sasdateform) {
                        x <- importConvertDateTime(x, 'date', 'sas')
                        changed <- TRUE
                    } else if (fi %in% sastimeform) {
                        x <- importConvertDateTime(x, 'time', 'sas')
                        changed <- TRUE
                    } else if (fi %in% sasdatetimeform) {
                        x <- importConvertDateTime(x, 'datetime', 'sas')
                        changed <- TRUE
                    } else if (force.integer) {
                        if (all(is.na(x))) {
                            storage.mode(x) <- 'integer'
                            changed <- TRUE
                        } else if (max(abs(x), na.rm = TRUE) <= (2 ^ 31 - 1) &&
                                   all(floor(x) == x, na.rm = TRUE)) {
                            storage.mode(x) <- 'integer'
                            changed <- TRUE
                        }
                    }
                } else if (possiblyConvertChar && is.character(x)) {
                    if ((is.logical(as.is) && !as.is) ||
                        (is.numeric(as.is) &&
                         length(unique(x)) < as.is * length(x))) {
                        x <- factor(x, exclude = '')
                        changed <- TRUE
                    }
                }
                
                lz <- lab[nam[i]]
                if (!is.null(lz) &&
                    length(lz) > 0 && !is.na(lz) && lz != '') {
                    names(lz) <- NULL
                    Hmisc::label(x)  <- lz
                    changed   <- TRUE
                }
                
                if (nam[i] %in% names(formats)  &&
                    formats[nam[i]] > "") {
                    SASformat(x) <- formats[[nam[i]]]
                    changed <- TRUE
                }
                
                if (nam[i] %in% names(iformats) &&
                    iformats[nam[i]] > "") {
                    SASformat(x) <- formats[[nam[i]]]
                    changed <- TRUE
                }
                
                if (changed) {
                    w[[i]] <- x
                }
            }
        }
        
        scat('.')
        
        res[[j]] <- w
    }
    
    scat("Done")
    
    if (include.formats) {
        nds <- nds + 1
        if (length(fds) > 0) {
            res$"FORMATS" <- ds[[fds]]
        } else {
            res$FORMATS <- empty.format.table()
        }
    }
    
    if (nds > 1 || as.list) {
        res
    } else {
        if (class(w) == "list") {
            w[[1]]
        } else {
            w
        }
    }
}